crates/ra_assists/src/handlers/change_lifetime_anon_to_named.rs

```diff
@@ -0,0 +1,143 @@
+use crate::{AssistContext, AssistId, Assists};
+use ra_syntax::{ast, ast::TypeParamsOwner, AstNode, SyntaxKind};
+use std::collections::HashSet;
+
+// Assist: change_lifetime_anon_to_named
+//
+// Change an anonymous lifetime to a named lifetime.
+//
+// ```
+// impl Cursor<'_<|>> {
+//     fn node(self) -> &SyntaxNode {
+//         match self {
+//             Cursor::Replace(node) | Cursor::Before(node) => node,
+//         }
+//     }
+// }
+// ```
+// ->
+// ```
+// impl<'a> Cursor<'a> {
+//     fn node(self) -> &SyntaxNode {
+//         match self {
+//             Cursor::Replace(node) | Cursor::Before(node) => node,
+//         }
+//     }
+// }
+// ```
+// FIXME: How can we handle renaming any one of multiple anonymous lifetimes?
+pub(crate) fn change_lifetime_anon_to_named(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let lifetime_token = ctx.find_token_at_offset(SyntaxKind::LIFETIME)?;
+    let lifetime_arg = ast::LifetimeArg::cast(lifetime_token.parent())?;
+    if lifetime_arg.syntax().text() != "'_" {
+        return None;
+    }
+    let next_token = lifetime_token.next_token()?;
+    if next_token.kind() != SyntaxKind::R_ANGLE {
+        // only allow naming the last anonymous lifetime
+        return None;
+    }
+    let impl_def = lifetime_arg.syntax().ancestors().find_map(ast::ImplDef::cast)?;
```

Comment:
I think this should be more generic and handle every node that can have lifetime parameters (basically, everhitng that implements `TypeParamsOwner`). I am ok with merging only support for impls, but having this working in every possible context would be significanlty more useful.

Reply:
i can see how it would be applicable to a function declaration, eg
```rust
fn my_fun(f: &Foo) -> &'_<|> Bar
```
to
```rust
fn my_fun(f: &'a Foo) -> &'a Bar
```

i don't see how it would be applicable to items such as struct. can you please advise? Thanks

Reply:
Ah, you are right of course, @jbalint, this doesn't work for structs (in fact, for structs we should have have a diagnostic instead). Yeah, I guess then we have only two cases here.

Reply:
I think i got all the `fn` cases

Reply:
@matklad can you clarify the comment about the diagnostic? is this something which we need to create another ticket for?